### PR TITLE
Make manifest backup/restore more robust, refactor to Ruby class and add tests

### LINKED ISSUE #123: backup_manifest fails when multiple manifest files are detected
Using `rails 3.2`, `capistrano 3.4.0` and `capistrano-rails 1.1.3`

Precompiling the assets works fine, but `backup_manifest` consistently fails due to having two manifest files (`manifest.json` and `manifest.yml`). When `detect_manifest_path` runs it will return a string as:

```ruby
"/home/deployer/releases/20150608160111/public/assets/manifest.json\n/home/deployer/releases/20150608160111/public/assets/manifest.yml"
```

The error occurs when `cp` attempts to move the files and fails due to the "\n" between the file paths. I updated the method in my local gem to use `gsub(/\s+/, ' ')` before the strip and the deployment worked:

```ruby
    def detect_manifest_path
      %w(
        .sprockets-manifest*
        manifest*.*
      ).each do |pattern|
        candidate = release_path.join('public', fetch(:assets_prefix), pattern)
        return capture(:ls, candidate).gsub(/\s+/, ' ').strip if test(:ls, candidate)
      end
      msg = 'Rails assets manifest file not found.'
      warn msg
      fail Capistrano::FileNotFound, msg
    end
```

### PR DESCRIPTION
**Pull request by mattbrictson**
This PR fixes #123 and #132. I've also taken this opportunity to add tests to capistrano-rails and enable Travis CI.

It is difficult to test Capistrano plugins without instantiating the entire Capistrano runtime; Capfile, deploy.rb, etc.; and most daunting of all, a Vagrant-based VM to SSH into. Hence the lack of tests in this project.

To address this, I've moved most of the manifest logic out of Rake tasks and into a pure Ruby class that can be unit tested.

As for the manifest logic itself, I've removed the dependency on `ls` to find manifests, which was fragile (parsing `ls` output is generally frowned upon) and known to fail in cases where more than one manifest file matched.

I also made the backup/restore system configurable, with new `:asset_manifest_backup_directory` and `:asset_manifest_patterns` settings.

@leehambley I would appreciate your honest feedback on this PR, specifically whether you think this is a good way to do testing and whether encapsulating logic into classes makes sense as a way to organize plugin code. Have I over-engineered this?

This is an alternative to the solutions proposed by @januszm @mzaccari in #124 and #133.

This PR needs the following before merging:

- [ ] CHANGELOG entry
- [ ] Handle missing backup scenario (`Capistrano::FileNotFound` logic)

### CODE REVIEW

**Comment by leehambley** [comment]
Thanks for the extensive PR @mattbrictson I'll need a bit more time to properly review this. I'll try and come back to you within the next 48 hours.

**Comment by mattbrictson** [comment]
@leehambley Thanks! Having slept on it, I think this PR has good intentions (make the code testable, move logic into Ruby class), but perhaps the wrong approach. I am desperate to add some sort of test coverage. But maybe this went too far in one direction.

Some things I am not happy with:

* Putting logic in a Ruby class at then accessing it from a Rake file using a memoized helper method is something that Capistrano's built-in SCMs do, too. But this means now code is spread across two files instead of one (and only the Ruby file is actually tested).
* Furthermore, the helper method (`def manifest` in my case, `def strategy` in the SCM case) is implicitly a global method added to `Object`, with can be dangerous (e.g. unintended name conflicts).
* The test code goes through great lengths to fake out Capistrano and SSHKit, which adds a lot of complexity.
* Test fixtures are tedious to create and are in a separate location from the test code, which can make the tests hard to follow. Perhaps a procedural way of setting up the tests would be better.

If we are serious about making Capistrano plugins more robust and testable, I think we should look to how Rake itself suggests it should be done. Maybe the right way to do it is to inherit from `Rake::TaskLib`, which seems the norm for how Rake tasks are distributed.

**Comment by leehambley** [comment]
> Furthermore, the helper method (def manifest in my case, def strategy in the SCM case) is implicitly a global method added to Object, with can be dangerous (e.g. unintended name conflicts).

True, but still we're *rarely* if ever loaded into *another application's* scope, so we're global for mostly ourselves, most of the time.

> The test code goes through great lengths to fake out Capistrano and SSHKit, which adds a lot of complexity.

This was the main reason I couldn't follow the code, and needed to postpone looking into it until Tuesday at least when I can cut out a block of time.

>  Maybe the right way to do it is to inherit from Rake::TaskLib, which seems the norm for how Rake tasks are distributed.

That's new to me. I'll read into it.

**Comment by mattbrictson** [comment]
Closing in favor of eventually making use of the plugin architecture proposed here: https://github.com/mattbrictson/experimental-capistrano-git-scm
